feat: support fargate spot

SoManyHs opened the pull request:
<!-- Provide summary of changes -->

Closes https://github.com/aws/copilot-cli/issues/2162

<!-- Issue number, if available. E.g. "Fixes #31", "Addresses #42, 77" -->

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

SoManyHs committed:
chore(template): add capacity providers

Also adds some logic in the stack package to ensure that autoscaling
configuration is not set if spot is specified for desiredCount.

SoManyHs committed:
chore: comment unused capacity provider for now

to appease the golint gods

SoManyHs committed:
feat(manifest): add Range subfields

Range can now either be a range string or a map containing the subfields
"min", "max" and "spot_from". This allows customers to use Application
Autoscaling with spot capacity.

Part of #2162

SoManyHs committed:
chore(stack): add range subfields to transformers

efekarakus commented on internal/pkg/deploy/cloudformation/stack/backend_svc.go:
```diff
@@ -83,10 +83,23 @@ func (s *BackendService) Template() (string, error) {
+
+	// desiredCountOnSpot and Autoscaling are mutually exclusive
+	var autoscaling *template.AutoscalingOpts
+	desiredCountOnSpot := s.manifest.Count.Autoscaling.Spot
```
Should we rename the `Autoscaling` type to `AdvancedCount`? since now it has more than just `range`
https://github.com/aws/copilot-cli/blob/6ec69950bc2d458ef9667284211fe1435c9b2f16/internal/pkg/manifest/svc.go#L58-L61

SoManyHs replied:
You know, I had actually tried changing `Autoscaling` into a type union called `AutoscalingOrSpot`, consisting of the `spot` count and then the existing Autoscaling struct but couldn't get the unmarshalling to work properly with the subfields ended up keeping it as it was.  I guess just changing the name would work too 😅

(see: https://github.com/SoManyHs/copilot-cli/commit/b054aae9380ab2b01aad0e3f4ebec28d35b82f69#diff-4465c235880a9abf21ad9b3cab6077493fa6f767be82b60324def2bbbc3b23e2)

efekarakus commented on internal/pkg/deploy/cloudformation/stack/lb_web_svc.go:
```diff
@@ -108,9 +108,22 @@ func (s *LoadBalancedWebService) Template() (string, error) {
 	}
-	autoscaling, err := convertAutoscaling(&s.manifest.Count.Autoscaling)
+
+	// desiredCountOnSpot and Autoscaling are mutually exclusive
```
What do you think of moving this logic to a separate function so that we can re-use it both for both type of services:
```go
type advancedCount struct {
  spot *int
  autoscaling *template.AutoScalingOpts
  cps *template.CapacityProviderStrategy
}

func convertAdvancedCount(config *manifest.AdvancedCount) (*advancedCount, error) { ... }
```

SoManyHs replied:
Good idea! Do you mean defined in the `stack` package, or would `advancedCount` also live in the `template` package?

efekarakus commented on internal/pkg/deploy/cloudformation/stack/lb_web_svc.go:
```diff
@@ -108,9 +108,22 @@ func (s *LoadBalancedWebService) Template() (string, error) {
+		var err error
+		autoscaling, err = convertAutoscaling(&s.manifest.Count.Autoscaling)
+		if err != nil {
+			return "", fmt.Errorf("convert the Auto Scaling configuration for service %s: %w", s.name, err)
```
nit: no need to wrap these err since it's calling private funcs

SoManyHs replied:
Oh, could you elaborate? I was following suit for the way the other workload fields are converted with errors being returned.

efekarakus replied:
meh it's not really important, i just thought that the error messages returned  from these `convert*` functions were already clear enough so we didn't need to wrap again (to follow [this](https://www.efekarakus.com/golang/2019/09/26/when-to-wrap-errors.html) although it's not applied everywhere 😛 )

efekarakus commented on internal/pkg/deploy/cloudformation/stack/transformers.go:
```diff
@@ -33,6 +33,12 @@ const (
+// Supported capacity providers for Fargate services
+const (
+	capacityProviderFargateSpot = "FARGATE_SPOT"
+	// capacityProviderFargate     = "FARGATE"
```
nit: can we remove dead comments and add them later when we need to initialize the field?

iamhopaul123 replied:
Maybe add it when it is in used. Just in case we never use it.

efekarakus commented on internal/pkg/deploy/cloudformation/stack/transformers.go:
```diff
@@ -45,6 +51,7 @@ var (
 	errUIDWithNonManagedFS = errors.New("UID and GID cannot be specified with non-managed EFS")
 	errInvalidUIDGIDConfig = errors.New("set managed filesystem access point creation info: must specify both UID and GID, or neither")
 	errReservedUID         = errors.New("set managed filesystem access point creation info: UID must not be 0")
+	errInvalidSpotConfig   = errors.New(`invalid spot configuration`)
```
```suggestion
	errInvalidSpotConfig   = errors.New(`"count.spot" and "count.range" cannot be specified together`)
```

iamhopaul123 replied:
Can we be more specific about the reason why the spot configuration is invalid? So it could be fixed more easily.

iamhopaul123 commented on internal/pkg/deploy/cloudformation/stack/transformers.go:
```diff
@@ -94,12 +101,41 @@ func parsePortMapping(s *string) (port *string, protocol *string, err error) {
+		Weight:           aws.Int(1),
+		CapacityProvider: capacityProviderFargateSpot,
+	}
+	cps = append(cps, spotCapacity)
```
```suggestion
	cps = append(cps, &template.CapacityProviderStrategy{
		Weight:           aws.Int(1),
		CapacityProvider: capacityProviderFargateSpot,
	})
```

iamhopaul123 commented on internal/pkg/deploy/cloudformation/stack/backend_svc.go:
```diff
@@ -83,10 +83,23 @@ func (s *BackendService) Template() (string, error) {
+	// desiredCountOnSpot and Autoscaling are mutually exclusive
+	var autoscaling *template.AutoscalingOpts
+	desiredCountOnSpot := s.manifest.Count.Autoscaling.Spot
+	if desiredCountOnSpot == nil {
```
Maybe we don't need this condition since it's already handled within `convertAutoscaling`?

efekarakus commented on internal/pkg/deploy/cloudformation/stack/workload.go:
```diff
@@ -97,13 +97,19 @@ func (w *wkld) StackName() string {
-		min, _, err := w.tc.Count.Autoscaling.Range.Parse()
-		if err != nil {
-			return nil, fmt.Errorf("parse task count value %s: %w", aws.StringValue((*string)(w.tc.Count.Autoscaling.Range)), err)
+		if w.tc.Count.Autoscaling.IgnoreRange() {
```
I wonder if we should move all this logic to a method on the `Count` type?
```go
func (c Count) Desired() (int, error) {
  if w.tc.Count.Value != nil { return ... }
  if w.tc.Count.Autoscaling.Spot != nil { return ... }
  // return min from autoscaling
}
```

SoManyHs replied:
Fix'd! https://github.com/aws/copilot-cli/pull/2188/commits/2cdef78bf8c607fb0f08c2aff59a7069fa4fa598

SoManyHs committed:
chore(template): add base field to capacity provider strategy

iamhopaul123 commented on internal/pkg/deploy/cloudformation/stack/transformers.go:
```diff
@@ -94,12 +101,41 @@ func parsePortMapping(s *string) (port *string, protocol *string, err error) {
 
+// convertCapacityProviders transforms the manifest fields into a format
+// parsable by the templates pkg.
+func convertCapacityProviders(a *manifest.Autoscaling) ([]*template.CapacityProviderStrategy, error) {
```
Since both `convertCapacityProviders` and `convertAutoscaling` take the same parameter, does it make more sense to combine them to just `convertAutoscaling` and return a struct with all info for the field in the manifest? It's a little strange to see a manifest field is parsed into three different fields. Ideally I think we should handle the `DesiredCountOnSpot` in the `TaskCount` in parameters, instead of using it in go template.

SoManyHs replied:
Yeah I totally get that it's weird! Unfortunately I wasn't sure how else to separate the arguments in a logical way, since range and count fields end up being part of the conversion logic. I also didn't feel like coupling the output data structure seemed like the right move, especially since `template.AutoscalingOpts` and Capacity  Provider Strategies aren't always used together and it seemed like having to handle nil checks every time for the combined data structure would result in more maintenance burden down the road.

iamhopaul123 replied:
Efe's proposal looks good to me https://github.com/aws/copilot-cli/pull/2188/files#r617104857, so that we'll just have `convertAdvancedCount` returning to the new struct that I mentioned before.

SoManyHs replied:
Fix'd! https://github.com/aws/copilot-cli/pull/2188/commits/bc88a31ff4b5cb5cd504c0ebf8e7b42bd17f48b4

iamhopaul123 commented on templates/workloads/partials/cf/service-base-properties.yml:
```diff
@@ -2,7 +2,9 @@ Cluster:
     !Sub '${AppName}-${EnvName}-ClusterId'
 TaskDefinition: !Ref TaskDefinition
-{{- if .Autoscaling }}
+{{- if .DesiredCountOnSpot}}
```
Maybe we don't need this condition?

SoManyHs replied:
We need this since if desiredCount is specified via the `spot` field, then we don't want the Autoscaling lambda to be part of the stack.

iamhopaul123 replied:
I might misunderstand something but I still feel this is a bit redundant since we already handle all the logic for desired count in https://github.com/aws/copilot-cli/pull/2188/files#diff-ac8fb9fb1f51e4365619b020fc6ccd1c32aab7ba83a8fce9e6a6e9ea7a341170R97 . The logic i think should be: if there's any autoscaling then we use the lambda value, if not, we use the `TaskCount`. The goal is I think we might not want the go template to handle too many logics.

SoManyHs committed:
chore: improve invalidSpotConfig error message

Co-authored-by: Efe Karakus <efekarakus@gmail.com>

SoManyHs committed:
refactor: append capacity provider strategy directly

Co-authored-by: Penghao He <penghaoh@amazon.com>

SoManyHs committed:
refactor: use convertAdvancedCount function

SoManyHs committed:
chore: rename manifest.Autoscaling to AdvancedCount

efekarakus commented on internal/pkg/deploy/cloudformation/stack/transformers.go:
```diff
@@ -94,12 +101,98 @@ func parsePortMapping(s *string) (port *string, protocol *string, err error) {
+		return nil, nil
+	}
+
+	spot := a.Spot
```
nit: we can omit this variable

efekarakus commented on internal/pkg/deploy/cloudformation/stack/transformers.go:
```diff
@@ -94,12 +101,98 @@ func parsePortMapping(s *string) (port *string, protocol *string, err error) {
+		Cps:         cps,
+	}
+
+	return ac, nil
```
nit:
```suggestion
	return &template.AdvancedCount{
		Spot:        a.Spot,
		Autoscaling: autoscaling,
		Cps:         cps,
	}, nil
```

efekarakus commented on internal/pkg/deploy/cloudformation/stack/backend_svc.go:
```diff
@@ -83,10 +83,22 @@ func (s *BackendService) Template() (string, error) {
+		return "", fmt.Errorf("convert the advanced count configuration for service %s: %w", s.name, err)
 	}
+
+	var autoscaling *template.AutoscalingOpts
+	var desiredCountOnSpot *int
+	var capacityProviders []*template.CapacityProviderStrategy
+
+	if advancedCount != nil {
```
What do you think of returning a concrete type from `convertAdvancedCount` so that we can skip this `nil` check:
```go
func convertAdvancedCount(a *manifest.AdvancedCount) (template.AdvancedCount, error)
```
then we can just do:
```go
template.WorkloadOpts{
   ...
   CapacityProviders: advancedCount.Cps,
   ...
```

SoManyHs replied:
I agree that the nil check isn't ideal, but I was wondering if having a potentially empty struct might not be potentially fiddly down the line? `convertAdvancedCount` is a tad odd bc it's just a utility function right now that abstracts away conversions that previously happened directly; in `stack/transformers.go` almost all the other conversion functions return pointer types. Just wanted to make sure we were okay breaking with that convention here?

SoManyHs committed:
refactor: move desiredCount logic onto manifest.Count

SoManyHs committed:
chore: fix assertion on require.Equal

Should be expected then actual: https://pkg.go.dev/github.com/stretchr/testify/require#Equal

efekarakus commented on internal/pkg/manifest/svc.go:
```diff
@@ -26,6 +27,50 @@ var ServiceTypes = []string{
 	BackendServiceType,
 }
 
+// RangeOpts contains either a Range or a range configuration for Autoscaling ranges
+type RangeOpts struct {
+	Range       *Range // Mutually exclusive with RangeConfig
+	RangeConfig RangeConfig
+}
```
What do you think of this to follow after what we did with the `Count` type?
```suggestion
// RangeOpts contains either a Range or a range configuration for Autoscaling ranges
type Range struct {
	Value       *IntRangeBand
	Advanced    AdvancedRange
}
```

SoManyHs replied:
https://github.com/aws/copilot-cli/pull/2188/commits/a4d886e3eb1a357a6668e1a3b1994bb991021b6a

efekarakus commented on internal/pkg/manifest/svc.go:
```diff
@@ -88,25 +149,49 @@ func (c *Count) UnmarshalYAML(unmarshal func(interface{}) error) error {
-// Autoscaling represents the configurable options for Auto Scaling as well as
+// Desired returns the desiredCount to be set on the CFN template
+func (c *Count) Desired() (*int, error) {
+	desiredCount := c.Value
```
nit: can we flip the conditionals to use guard-statements to reduce nesting?
```go
if c.AdvancedCount.IsEmpty() {
   return c.Value
}
if c.AdvancedCount.IgnoreRange() {
  return c.AdvancedCount.Spot
}
...
```

SoManyHs replied:
https://github.com/aws/copilot-cli/pull/2188/commits/f6362c96507dd91f2b2bed70c8b93ae26f2cba73

efekarakus reviewed (commented):
🎉 Looks great! Just some small changes 🙏 and then 🚀

SoManyHs committed:
refactor: return Advanced count directly w/o variable assignment

Co-authored-by: Efe Karakus <efekarakus@gmail.com>

SoManyHs committed:
chore: remove unused variable

SoManyHs committed:
chore: rename Range

Rename custom types: RangeOpts -> Range, Range -> IntRangeBand

SoManyHs committed:
refactor: reduce nesting in Count.Desired()

SoManyHs committed:
Merge branch 'mainline' into fargate-spot

iamhopaul123 reviewed (approved):
ship it :shipit:
